src/wc/wc.rs

```diff
@@ -133,6 +160,31 @@ fn is_word_seperator(byte: u8) -> bool {
     byte == SPACE || byte == TAB || byte == CR || byte == SYN || byte == FF
 }
 
+#[inline]
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn count_bytes_using_splice(fd: RawFd) -> nix::Result<usize> {
+    const BUF_SIZE: usize = 16 * 1024;
+
+    let null_file = OpenOptions::new()
+        .write(true)
+        .open("/dev/null")
+        .map_err(|_| nix::Error::last())?;
+    let null = null_file.as_raw_fd();
+    let (pipe_rd, pipe_wr) = pipe()?;
+
+    let mut byte_count = 0;
+    loop {
+        let res = splice(fd, None, pipe_wr, None, BUF_SIZE, SpliceFFlags::empty())?;
```

Comment:
So, I think this could be turned into something for the collection of zero-copy stuff in `uucore`.  I think there could actually be two types created from this code.

The first would be something like a reader that can perform zero-copy reads when used by other zero-copy types, but otherwise function as a normal reader for normal Rust code (_e.g._ if you used the reader to read into a `String`, it would act like a normal `File` reader or whatever).  Maybe the zero-copy read method just returns some `struct` representing a platform-independent pipe, which would then need to implement `Read` and `Write` as well.

The second type that I think can be created takes the zero-copy reader I just mentioned and the zero-copy writer that already exists (but which would probably need to be extended) and basically reads the data from the reader and writes it out into the writer without any copies.  It could either take the zero-copy reader and writer as arguments or construct them internally.

Sorry if this isn't clear, I'm just kind of thinking out loud.  Let me know if I need to clarify my thoughts.

Reply:
I think your idea of creating a cross-platform abstraction over the raw zero-copy Linux system calls is a good one, but I think it belongs as a seperate crate, so that other projects can benefit as well.

That said, I have the following reservation: I don't have a clear idea of how a general API would look like — and the fact that these functions are so seldom used doesn't help. Could it be that these functions are specialized enough accessing them through general interface would result in a performance hit? I'd have to research it, but I'm not sure, for example, if it makes sense to use `vmsplice()` if you're not calling it in a tight loop such as in `yes.rs` (GitHub code search maybe?).

Thus if it were only me, I would first implement the "zero-copyable" parts of coreutils using the raw system calls such as I have been doing, and then, if I see a common trend of how they are used, publish a crate with an abstraction for that use case. However, it's not only me — you seem to have an idea of how such an API could look like. I haven't yet fully digested [what you have already commited](https://github.com/uutils/uucore/tree/c6cacb0f6a62a9aadcee4dfdd7d57a179cddf544/src/zero_copy) to `uucore`, but even so I don't think I have enough Rust or Linux experience to be a good judge of such an API currently.

Reply:
If we were to make abstractions for them, I think we would end up using them more.  Unless `vmsplice()`/`splice()` hurt performance themselves, I don't think a general interface will be an issue as on non-Linux systems it should optimize to just `read()`/`write()`, and on Linux the overhead should be insignificant compared to the actual read/write.

I am planning on splitting off much of the functionality in `uucore` into separate crates, but at the moment I think it can serve as a sort of testing ground for functionality.  I'll see if I can implement what I was talking about in `uucore`.  In the meantime, using the raw system calls is fine I guess.  I'd like it if you would create an issue indicating which utilities are using the raw system calls though (so I won't need to track them down later).